cirq-ft/cirq_ft/algos/arithmetic_gates.py

```diff
@@ -161,9 +222,104 @@ def __pow__(self, power: int):
+        # into `prefix_equality` using d = |len(P) - len(Q)| And operations => 4d T.
+        if len(P) != len(Q):
+            [prefix_equality] = context.qubit_manager.qalloc(1)
+            if len(P) > len(Q):
+                decomposition = tuple(
+                    cirq.flatten_to_ops(
+                        _equality_with_zero(context, tuple(P[:-n]), prefix_equality)
+                    )
+                )
+                yield from decomposition
+                adjoint.extend(cirq.inverse(op) for op in decomposition)
+            else:
+                decomposition = tuple(
+                    cirq.flatten_to_ops(
+                        _equality_with_zero(context, tuple(Q[:-n]), prefix_equality)
+                    )
+                )
+                yield from decomposition
+                adjoint.extend(cirq.inverse(op) for op in decomposition)
+
+                yield cirq.X(target), cirq.CNOT(prefix_equality, target)
+
+        # compare the remaing suffix of P and Q
+        P = P[-n:]
+        Q = Q[-n:]
+        decomposition = tuple(
+            cirq.flatten_to_ops(self._decompose_via_tree(context, tuple(P), tuple(Q)))
+        )
+        adjoint.extend(cirq.inverse(op) for op in decomposition)
+        yield from decomposition
+
+        less_than, greater_than = context.qubit_manager.qalloc(2)
+        decomposition = tuple(
+            cirq.flatten_to_ops(compare_qubits(P[-1], Q[-1], less_than, greater_than))
+        )
+        adjoint.extend(cirq.inverse(op) for op in decomposition)
+        yield from decomposition
+
+        if prefix_equality is None:
+            yield cirq.X(target)
+            yield cirq.CNOT(greater_than, target)
+        else:
+            [less_than_or_equal] = context.qubit_manager.qalloc(1)
+            yield and_gate.And([1, 0]).on(prefix_equality, greater_than, less_than_or_equal)
+            adjoint.append(
+                and_gate.And([1, 0], adjoint=True).on(
+                    prefix_equality, greater_than, less_than_or_equal
+                )
+            )
+
+            yield cirq.CNOT(less_than_or_equal, target)
+
+        yield from reversed(adjoint)
+
     def _t_complexity_(self) -> infra.TComplexity:
-        # TODO(#112): This is rough cost that ignores cliffords.
-        return infra.TComplexity(t=4 * (self.x_bitsize + self.y_bitsize))
+        n = min(self.x_bitsize, self.y_bitsize)
+        d = max(self.x_bitsize, self.y_bitsize) - n
+        is_second_longer = self.y_bitsize > self.x_bitsize
+        if d == 0:
+            return infra.TComplexity(t=8 * n - 4, clifford=46 * n - 17)
+        elif d == 1:
+            return infra.TComplexity(t=8 * n, clifford=46 * n + 3 + 2 * is_second_longer)
+        else:
+            return infra.TComplexity(
+                t=8 * n + 4 * d - 4, clifford=46 * n + 17 * d - 14 + 2 * is_second_longer
+            )
+
+    def _has_unitary_(self):
+        return True
```

Comment:
why do we need to override this? shouldn't it figure it out from the fact that we have a decomposition?

Reply:
bump

Reply:
the simulation tests fail if we don't. the simulators check that what it simulates is unitary with some measurements.

the cirq.has_unitary now fails on the decomposition because the decomposition uses non-unitary operations (all the And(adjoint=True)) but we don't have support for that check now except by simulation